File: core/evolution_runtime/runtime_memory_engine.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict

from core.crypto.kaalka_runtime_engine import decrypt_value, encrypt_value


def save_evolution_runtime(
    path: str,
    memory: Dict[str, Any],
    key: str,
) -> Dict[str, Any]:
    payload = json.dumps(memory, sort_keys=True)
    encrypted = encrypt_value(payload, key)
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(
        json.dumps({"encrypted": encrypted["encrypted"], "algorithm": "kaalka"}, sort_keys=True),
        encoding="utf-8",
    )
    return {"saved": True, "path": str(target), "algorithm": "kaalka", "bounded": True}


def load_evolution_runtime(
    path: str,
    key: str,
) -> Dict[str, Any]:
    target = Path(path)
    if not target.exists():
        return {"available": False, "memory": _empty_memory(), "bounded": True}

    wrapper = json.loads(target.read_text(encoding="utf-8"))
    decrypted = decrypt_value(wrapper["encrypted"], key)
    memory = json.loads(decrypted["decrypted"])

    return {
        "available": True,
        "memory": memory,
        "algorithm": "kaalka",
        "bounded": True,
    }


def remember_evolution_runtime(
    memory: Dict[str, Any],
    update: Dict[str, Any],
) -> Dict[str, Any]:
    merged = dict(memory)
    for field in (
        "evolution_histories",
        "selector_lineage",
        "workflow_evolution",
        "semantic_evolution",
        "optimization_histories",
    ):
        merged.setdefault(field, update.get(field, merged.get(field, {})))
    merged.update(update)
    merged["bounded"] = True
    return merged


def _empty_memory() -> Dict[str, Any]:
    return {
        "evolution_histories": [],
        "selector_lineage": [],
        "workflow_evolution": {},
        "semantic_evolution": {},
        "optimization_histories": [],
        "bounded": True,
    }

File: src/evolution_runtime/runtimeMemoryEngine.ts
/**
 * Converted from Python: core/evolution_runtime/runtime_memory_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { decryptValue, encryptValue } from "../crypto/kaalkaRuntimeEngine.js";

export function saveEvolutionRuntime(path: any, memory: any, key: any): any {
  var payload: any = py.jsonDumps(memory, {sortKeys: true});
  var encrypted: any = encryptValue(payload, key);
  var target: any = py.path(path);
  target.parent.mkdir(true, true);
  target.write_text(py.jsonDumps({"encrypted": py.at(encrypted, "encrypted"), "algorithm": "kaalka"}, {sortKeys: true}), "utf-8");
  return {"saved": true, "path": py.toStr(target), "algorithm": "kaalka", "bounded": true};
}
export function loadEvolutionRuntime(path: any, key: any): any {
  var target: any = py.path(path);
  if (!py.truthy(target.exists())) {
    return {"available": false, "memory": _emptyMemory(), "bounded": true};
  }
  var wrapper: any = py.jsonLoads(target.read_text("utf-8"));
  var decrypted: any = decryptValue(py.at(wrapper, "encrypted"), key);
  var memory: any = py.jsonLoads(py.at(decrypted, "decrypted"));
  return {"available": true, "memory": memory, "algorithm": "kaalka", "bounded": true};
}
export function rememberEvolutionRuntime(memory: any, update: any): any {
  var merged: any = py.pyDict(memory);
  var field: any;
  for (field of py.iter(["evolution_histories", "selector_lineage", "workflow_evolution", "semantic_evolution", "optimization_histories"])) {
    py.setdefault(merged, field, py.get(update, field, py.get(merged, field, {})));
  }
  py.update(merged, update);
  py.setItem(merged, "bounded", true);
  return merged;
}
export function _emptyMemory(): any {
  return {"evolution_histories": [], "selector_lineage": [], "workflow_evolution": {}, "semantic_evolution": {}, "optimization_histories": [], "bounded": true};
}
export { decryptValue, encryptValue };
